File: engine/tools/lead_scoring.py
"""Lead scoring 0–100 for outreach prioritisation."""

from __future__ import annotations

from typing import Any

_PREFERRED_LOCAL = ("info@", "contact@", "hello@", "enquiries@", "office@")


def _contact_quality_score(email: str) -> int:
    low = (email or "").lower()
    if any(low.startswith(p) for p in _PREFERRED_LOCAL):
        return 15
    if "@" in low:
        return 8
    return 0


def _website_quality_score(research: dict[str, Any]) -> int:
    score = 0
    if research.get("has_https"):
        score += 10
    if research.get("has_contact_page"):
        score += 8
    text_len = int(research.get("page_text_length") or 0)
    if text_len > 1500:
        score += 7
    elif text_len > 400:
        score += 4
    return min(25, score)


def _industry_fit_score(sector: str, campaign_id: str) -> int:
    sector = (sector or "generic").lower()
    if campaign_id == "weathers":
        fit_sectors = {
            "restaurant", "hotel", "care_home", "school", "letting_agent",
            "pub", "gym", "pet_groomer", "bakery", "food_production",
        }
        return 25 if sector in fit_sectors else 12 if sector != "generic" else 8
    if campaign_id == "pesttrace":
        return 25 if sector in ("pest_control_firm", "generic") else 15
    return 12


def _company_size_score(research: dict[str, Any], phone: str) -> int:
    score = 0
    services = research.get("services") or []
    if isinstance(services, list) and len(services) >= 2:
        score += 10
    if (phone or "").strip() or str(research.get("phone") or "").strip():
        score += 5
    text_len = int(research.get("page_text_length") or 0)
    if text_len > 3000:
        score += 5
    return min(20, score)


def _research_boost_score(research: dict[str, Any], page_text: str = "") -> int:
    score = 0
    if str(research.get("contact_name") or "").strip():
        score += 10
    if str(research.get("phone") or "").strip():
        score += 5
    blob = (page_text or "").lower()
    if any(k in blob for k in ("compliance", "audit", "haccp", "food safety", "ipc")):
        score += 10
    return min(25, score)


def _local_market_score(country: str, city: str, campaign_id: str) -> int:
    country = (country or "").upper()
    if campaign_id == "weathers":
        return 15 if country == "UK" and city else 10 if country == "UK" else 0
    target = {"UK", "IE", "DE", "FR", "ES", "IT", "NL", "IN", "US", "CA", "AU"}
    return 15 if country in target else 5


def compute_lead_score(prospect: dict[str, Any], research: dict[str, Any] | None = None) -> tuple[int, dict[str, int]]:
    """Return (score 0–100, breakdown)."""
    raw = prospect.get("raw") or {}
    if not isinstance(raw, dict):
        raw = {}
    research = research or raw.get("research") or {}
    if not isinstance(research, dict):
        research = {}

    campaign = str(prospect.get("campaign") or "pesttrace").lower()
    sector = str(prospect.get("sector") or research.get("sector") or "generic")
    email = str(prospect.get("email") or "")
    country = str(prospect.get("country") or "")
    city = str(prospect.get("city") or "")
    phone = str(prospect.get("phone") or "")

    page_text = str(research.get("page_text_sample") or "")
    breakdown = {
        "website_quality": _website_quality_score(research),
        "industry_fit": _industry_fit_score(sector, campaign),
        "company_size": _company_size_score(research, phone),
        "contact_quality": _contact_quality_score(email),
        "local_market": _local_market_score(country, city, campaign),
        "research_boost": _research_boost_score(research, page_text),
    }
    total = min(100, sum(breakdown.values()))
    return total, breakdown


def persist_lead_score(prospect_id: str, score: int, breakdown: dict[str, int]) -> None:
    from supabase_client import get_supabase

    try:
        sb = get_supabase()
        row = (
            sb.table("outreach_prospects")
            .select("raw")
            .eq("id", prospect_id)
            .single()
            .execute()
        )
        raw = row.data.get("raw") if row.data else {}
        if not isinstance(raw, dict):
            raw = {}
        raw["score"] = breakdown
        sb.table("outreach_prospects").update(
            {"lead_score": score, "raw": raw}
        ).eq("id", prospect_id).execute()
    except Exception:
        pass

File: web/src/lib/outreach/lead-scoring.ts
const PREFERRED_LOCAL = ["info@", "contact@", "hello@", "enquiries@", "office@"];

type Research = {
  services?: string[];
  has_https?: boolean;
  has_contact_page?: boolean;
  page_text_length?: number;
  page_text_sample?: string;
  sector?: string;
  contact_name?: string;
  phone?: string;
};

function contactQualityScore(email: string): number {
  const low = email.toLowerCase();
  if (PREFERRED_LOCAL.some((p) => low.startsWith(p))) return 15;
  if (low.includes("@")) return 8;
  return 0;
}

function websiteQualityScore(research: Research): number {
  let score = 0;
  if (research.has_https) score += 10;
  if (research.has_contact_page) score += 8;
  const len = research.page_text_length ?? 0;
  if (len > 1500) score += 7;
  else if (len > 400) score += 4;
  return Math.min(25, score);
}

function industryFitScore(sector: string, campaignId: string): number {
  const s = sector.toLowerCase();
  if (campaignId === "weathers") {
    const fit = new Set([
      "restaurant", "hotel", "care_home", "school", "letting_agent",
      "pub", "gym", "pet_groomer", "bakery", "food_production",
    ]);
    if (fit.has(s)) return 25;
    return s !== "generic" ? 12 : 8;
  }
  if (campaignId === "pesttrace") {
    return s === "pest_control_firm" || s === "generic" ? 25 : 15;
  }
  if (campaignId === "jgdevs") {
    const fit = new Set([
      "tradesperson", "salon", "local_shop", "professional",
      "restaurant", "gym", "pet_groomer", "pub", "bakery", "generic",
    ]);
    return fit.has(s) ? 25 : 12;
  }
  return 12;
}

function companySizeScore(research: Research, phone: string): number {
  let score = 0;
  if ((research.services?.length ?? 0) >= 2) score += 10;
  if (phone.trim()) score += 5;
  if ((research.page_text_length ?? 0) > 3000) score += 5;
  return Math.min(20, score);
}

function researchBoostScore(research: Research): number {
  let score = 0;
  if (String(research.contact_name || "").trim()) score += 10;
  if (String(research.phone || "").trim()) score += 5;
  const blob = (research.page_text_sample || "").toLowerCase();
  if (["compliance", "audit", "haccp", "food safety", "ipc"].some((k) => blob.includes(k))) {
    score += 10;
  }
  return Math.min(25, score);
}

function localMarketScore(country: string, city: string, campaignId: string): number {
  const c = country.toUpperCase();
  if (campaignId === "weathers") {
    return c === "UK" && city ? 15 : c === "UK" ? 10 : 0;
  }
  if (campaignId === "jgdevs") {
    const eu = new Set(["UK", "IE", "DE", "FR", "ES", "IT", "NL", "BE", "AT", "PT", "PL", "SE", "DK"]);
    return eu.has(c) && city ? 15 : eu.has(c) ? 10 : 0;
  }
  const target = new Set(["UK", "IE", "DE", "FR", "ES", "IT", "NL", "IN", "US", "CA", "AU"]);
  return target.has(c) ? 15 : 5;
}

export function computeLeadScore(prospect: {
  campaign?: string | null;
  sector?: string | null;
  email?: string | null;
  country?: string | null;
  city?: string | null;
  phone?: string | null;
  raw?: { research?: Research } | null;
}): { score: number; breakdown: Record<string, number> } {
  const research = prospect.raw?.research ?? {};
  const campaign = (prospect.campaign || "pesttrace").toLowerCase();
  const sector = (prospect.sector || research.sector || "generic").toString();

  const breakdown = {
    website_quality: websiteQualityScore(research),
    industry_fit: industryFitScore(sector, campaign),
    company_size: companySizeScore(research, prospect.phone || research.phone || ""),
    contact_quality: contactQualityScore(prospect.email || ""),
    local_market: localMarketScore(prospect.country || "", prospect.city || "", campaign),
    research_boost: researchBoostScore(research),
  };

  if (campaign === "jgdevs") {
    let opp = 0;
    if (!research.has_https) opp += 10;
    const len = research.page_text_length ?? 0;
    if (len < 500) opp += 12;
    if (!research.has_contact_page) opp += 8;
    const blob = (research.page_text_sample || "").toLowerCase();
    if (!["book", "booking", "appointment", "schedule"].some((k) => blob.includes(k))) opp += 5;
    breakdown.website_quality = Math.min(25, opp);
  }

  const score = Math.min(100, Object.values(breakdown).reduce((a, b) => a + b, 0));
  return { score, breakdown };
}
